### ISSUE TITLE
Support for clang's `noescape` attribute

### ISSUE THREAD

**Issue by nemequ** [nemequ/hedley#12]
Clang recently added a [`noescape`](https://clang.llvm.org/docs/AttributeReference.html#noescape) attribute that might be worth a Hedley macro.  I don't think other compilers have anything similar, but still useful.

**Comment by travisdowns** [comment]
Indeed, I guess the minimum bar for inclusion into Hedly is "at least 1 compiler implements it", right?

Because as soon as one compiler implements it, you may want to use it and you want hedley for portability even though all the other implementations are empty.

**Comment by nemequ** [comment]
Yeah, one compiler **or static analyzer** is definitely enough.

I'd sort of forgotten about this issue.  If you think it's useful I can go ahead and add it to the next release (probably early next week); the implementation should be pretty straightforward since AFAIK only clang supports it.  Even if clang doesn't take full advantage today, there should be plenty of opportunities to exploit in the future for this, especially when the compiler doesn't have visibility into the implementation.

**Comment by travisdowns** [comment]
Heh, I didn't notice I was replying on an old issue: I came for something else and just saw this.

I don't remember if I learned about `noescape` here originally, but in the last year I haven't seen much of it, but I think it can be a big deal for some optimizations. Basically I guess it can enable both heap allocation elimination and avoid some aliasing problems (since if a compiler knows a pointer did not escape it knows the "doesn't alias" state is preserved across the function call). I haven't checked whether either of these actually occur on current clang though.
